samples/Config.CustomConfigurationProviders.Sample/Microsoft.Extensions.Configuration.ConfigFile/ConfigFileConfigurationProvider.cs

```diff
@@ -0,0 +1,147 @@
+    ///     &lt;/appSettings&gt;
+    ///     &lt;connectionStrings&gt;
+    ///       &lt;add name=&quot;MyJetConn&quot; connectionString=&quot;Provider=Microsoft.Jet.OLEDB.4.0; Data Source=C:\\testdatasource.accdb; Persist Security Info=False;&quot; providerName=&quot;System.Data.OleDb&quot; /&gt;
+    ///       &lt;add name=&quot;MyExcelConn&quot; connectionString=&quot;Dsn=Excel Files;dbq=data.xlsx;defaultdir=.; driverid=790;maxbuffersize=2048;pagetimeout=5&quot; providerName=&quot;System.Data.Odbc&quot; /&gt;
+    ///     &lt;/connectionStrings&gt;
+    ///     &lt;sampleSection&gt;
+    ///       &lt;add key=&quot;setting1&quot; value=&quot;This is the setting1 value&quot; /&gt;
+    ///       &lt;add key=&quot;setting2&quot; value=&quot;This is the setting2 value&quot; /&gt;
+    ///     &lt;/sampleSection&gt;
+    ///   &lt;/configuration&gt;
+    ///
+    /// Code:
+    ///   var configuration = new ConfigurationBuilder().AddConfigFile(&quot;Web.config&quot;).Build();
+    ///   var setting1 = configuration[&quot;sampleSection:setting1&quot;];
+    ///   var allAppSettings = configuration.GetSection(&quot;appSettings&quot;).GetChildren();
+    ///   var myJetConnection = configuration[&quot;connectionStrings:MyJetConn&quot;];
+    /// </example>
+    public class ConfigFileConfigurationProvider : ConfigurationProvider
+    {
+        private readonly ILogger _logger;
+        private readonly string _configuration;
+        private readonly bool _loadFromFile;
+        private readonly bool _isOptional;
+
+        private readonly IEnumerable<IConfigurationParser> _parsers;
+
+        public ConfigFileConfigurationProvider(string configuration, bool loadFromFile, bool optional, params IConfigurationParser[] parsers)
+            : this(configuration, loadFromFile, optional, null, parsers)
+        { }
+
+        public ConfigFileConfigurationProvider(string configuration, bool loadFromFile, bool optional, ILogger logger, params IConfigurationParser[] parsers)
+        {
+            _loadFromFile = loadFromFile;
+            _configuration = configuration;
+            _isOptional = optional;
+            _logger = logger;
+
+            var parsersToUse = new List<IConfigurationParser> {
+                new KeyValueParser(),
+                new KeyValueParser("name", "connectionString")
+            };
+
+            parsersToUse.AddRange(parsers);
+
+            _parsers = parsersToUse.ToArray();
+        }
+
+        public override void Load()
+        {
+            if (_loadFromFile && !_isOptional && !File.Exists(_configuration))
+            {
+                throw new FileNotFoundException("Could not find configuration file to load!", _configuration);
+            }
+
+            XDocument document;
+
+            if (!TryLoadXDocument(out document))
+            {
+                return;
+            }
+
+            Data = new SortedDictionary<string, string>(StringComparer.OrdinalIgnoreCase);
```

Comment:
Also, the `Data` property is set here to an empty dictionary, and then a few lines below set again to the real dictionary with actual data. Can probably skip this line here.